infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/LogReplicationAckReader.java

```diff
@@ -111,43 +122,169 @@ private long getMaxReplicatedStreamsTail() {
+    /**
+     * The calculation of remaining entries during log entry sync takes into account:
+     * - lastAckedTimestamp
+     * - txStreamTail
+     * - currentTxStreamProcessedTs
+     *
+     * Consider the following representation of the data log, where entries 20, 50, 60, 70 belong to the
+     * tx stream.
+     *
+
+     +---+---+---+---+---+---+---+---+---+---+---+---+
+     |   |   |   |   |   |   |   |   |   |   |   |   |
+     +---+---+---+---+---+---+---+---+---+---+---+---+
+     20          50          60       70         100
+     tx          tx          tx       tx        (log tail / not part of tx stream)
+
+
+     * Case 1.0: Log Entry Sync lagging behind (in processing) current processing not acked with entries to replicate
+     *          - lastAckedTimestamp = 50
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 60, true (contains replicated streams)
+     *
+     *          remainingEntries = entriesBetween(50, 70] = 2
+
+     * Case 1.1: Log Entry Sync lagging behind (in processing) current processing no entries to replicate
+     *          - lastAckedTimestamp = 50
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 60, false (does not contain streams to replicate)
+     *
+     *          remainingEntries = entriesBetween(60, 70] = 1
+
+     * Case 2.0: Log Entry Sync Up to Date
+     *          - lastAckedTimestamp = 70
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 70, true (contains replicated streams)
+     *
+     *          remainingEntries = 0
+
+     * Case 2.1: Log Entry Sync Up to Date
+     *          - lastAckedTimestamp = 60
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 70, false (does not contain replicated streams,
+     *          so there is no expectation that lastAckedTimestamp reaches 70)
+     *
+     *          remainingEntries = 0
+
+
+     * Case 2.2: Log Entry Sync Lagging Behind (in ack)
+     *          - lastAckedTimestamp = 60
+     *          - txStreamTail = 70
+     *          - currentTxStreamProcessedTs = 70, true (does contain replicated streams,
+     *           wait until lastAckedTimestamp reflects this)
+     *
+     *          remainingEntries = entriesBetween(60, 70] = 1
+     *
+     */
+    private long calculateRemainingEntriesIncrementalUpdates(long lastAckedTs, long txStreamTail,
+                                                             StreamIteratorMetadata currentTxStreamProcessedTs) {
+        long noRemainingEntriesToSend = 0;
+
+        // If we have processed up to or beyond the latest known tx stream tail
+        // we can assume we are up to date.
+        // Note: in the case of a switchover the tail won't be moving so we can assume
+        // the tx's stream last processed timestamp will never be above the stream's tail,
+        // but in the case of ongoing replication or holes, it might be above.
```

Comment:
in case of holes, Tx stream's tail will not be above the stream tail, right?  It will be the other way round

Reply:
it could right, imagine by the time we query the tail it's 10.. when we check lastProcessedEntry we see 12... and 12 is a hole.. cause on non-switchover cases the tail can continue growing but we freezed it at a certain point.